### DISCUSSION TITLE
OIDC and Google Cloud Storage

### DISCUSSION THREAD

**Discussion by kolban-google** [drakkan/sftpgo#1024] [Q&A]
If I follow, I can use OIDC to sign in to SFTPGO using my OIDC provider ... eg. Google.  However, imagine I'm using Google Cloud Storage as my file system.  When I define Google Cloud Storage, I have to provide a credentials file that is then used to define my Google Cloud Storage access.

This then begs the question ... what is the relation between OIDC and Google Cloud Storage?

What I want to achieve is:

1. Login to the web interface of SFTPGO without having previously defined a userid at SFTPGO
2. SFTPGO will authenticate me based on OIDC as an identity that has permissions to access Google Cloud Storage
3. Use SFTPGO to work with Google Cloud Storage files.

While I can see how SFTGO uses OIDC to authenticate me, I am not seeing a relationship between the authenication of that user and the authorizations that the user has with Google Cloud Storage.

**Comment by drakkan** [comment]
> If I follow, I can use OIDC to sign in to SFTPGO using my OIDC provider ... eg. Google. However, imagine I'm using Google Cloud Storage as my file system. When I define Google Cloud Storage, I have to provide a credentials file that is then used to define my Google Cloud Storage access.
> 
> This then begs the question ... what is the relation between OIDC and Google Cloud Storage?
> 
> What I want to achieve is:
> 
> 1. Login to the web interface of SFTPGO without having previously defined a userid at SFTPGO
> 2. SFTPGO will authenticate me based on OIDC as an identity that has permissions to access Google Cloud Storage
> 3. Use SFTPGO to work with Google Cloud Storage files.
> 
> While I can see how SFTGO uses OIDC to authenticate me, I am not seeing a relationship between the authenication of that user and the authorizations that the user has with Google Cloud Storage.

There is no relation. You can pre-create users in SFTPGo and authenticate them via OIDC or you can create users on the fly (just after the OIDC authentication) using a pre-login hook. The [custom_fields](https://github.com/drakkan/sftpgo/blob/main/sftpgo.json#L282) configuration parameter can be used to define OIDC claim field names to pass to the pre-login hook, these fields can be used for implementing custom logic when creating/updating the SFTPGo users within the hook.

If you have a single bucket, you can configure the bucket credentials via [env vars](https://cloud.google.com/docs/authentication/application-default-credentials#GAC) and use "automatic credentials" for SFTPGo users. When you create users in your pre-login hook, you can restrict them to a bucket prefix, for example `users/<username>/`. This way each user will only use a portion of your bucket. A such pre-login hook should be quite simple to write
